jina/types/document/multimodal.py

```diff
@@ -0,0 +1,108 @@
+
+from . import Document
+from ..sets.chunk_set import ChunkSet
+from ...proto import jina_pb2
+from ..ndarray.generic import NdArray
+from ...excepts import LengthMismatchException, BadDocType
+
+__all__ = ['MultimodalDocument', 'DocumentContentType']
+
+DocumentContentType = TypeVar('DocumentContentType', bytes, str,
+                              np.ndarray, jina_pb2.NdArrayProto, NdArray)
+
+class MultimodalDocument(Document):
+    """
+    :class:`MultimodalDocument` is a data type created based on Jina primitive data type :class:`Document`.
+
+    It shares the same methods and properties with :class:`Document`, while it focus on modality at chunk level.
+
+    .. warning::
+        - It assumes that every ``chunk`` of a ``document`` belongs to a different modality.
+        - It assumes that every :class:`MultimodalDocument` have at least two chunks.
+    """
+    def __init__(self, document = None, chunks: List[Document] = None, copy: bool = False, **kwargs):
+        """
+
+        :param document: the document to construct from. If ``bytes`` is given
+                then deserialize a :class:`DocumentProto`; ``dict`` is given then
+                parse a :class:`DocumentProto` from it; ``str`` is given, then consider
+                it as a JSON string and parse a :class:`DocumentProto` from it; finally,
+                one can also give `DocumentProto` directly, then depending on the ``copy``,
+                it builds a view or a copy from it.
+        :param chunks: the chunks of the multimodal document to initialize with. Expected to
+                received a list of :class:`Document`, with different modalities.
+        :param copy: when ``document`` is given as a :class:`DocumentProto` object, build a
+                view (i.e. weak reference) from it or a deep copy from it.
+        :param kwargs: other parameters to be set
+        """
+        super().__init__(document=document, copy=copy, **kwargs)
+        self._modality_content_mapping = {}
+        if chunks:
+            self._validate(chunks)
+            self.chunks.extend(chunks)
+            self._handle_chunk_attributes()
+
+    def _build_modality_content_mapping(self) -> Dict:
+        for chunk in self.chunks:
+            modality = chunk.modality
+            self._modality_content_mapping[modality] = chunk.embedding \
+                if chunk.embedding is not None \
+                else chunk.content
+        self._validate(chunks=self.chunks)
+
+    def _validate(self, chunks: List[Document]):
+        modalities = set([chunk.modality for chunk in chunks])
+        if len(chunks) < 2:
+            raise BadDocType('MultimodalDocument should consist at least 2 chunks.')
+        if len(modalities) != len(chunks):
+            raise LengthMismatchException(f'Length of modality is not identical to length of chunks.')
+
+    def _handle_chunk_attributes(self):
+        """Handle chunk attributes, such as :attr:`granularity` and :attr:`mime_type`.
```

Comment:
What if the chunks come already with the proper granularity but the document has None?

Reply:
We should at lesst add a warning in the constructor about how the granularity behaves. Would not change it but sure add docstring

Reply:
@JoanFM I did check it out, the document will have a default `granularity ` as 0.
![image](https://user-images.githubusercontent.com/9794489/100805567-2bed2080-342f-11eb-8dfb-a4da8dea5516.png)

If you take a look at our code logic in `DocumentSet` or `MatchSet`, you'll see something similar:
```python
    def append(self, document: Optional['Document'] = None, **kwargs) -> 'Document':
        """Add a sub-document (i.e chunk) to the current Document

        :return: the newly added sub-document in :class:`Document` view

        .. note::
            Comparing to :attr:`DocumentSet.append()`, this method adds more safeguard to
            make sure the added chunk is legit.
        """
        c = self._docs_proto.add()
        if document is not None:
            c.CopyFrom(document.as_pb_object)

        from ..document import Document
        with Document(c) as chunk:
            chunk.set_attrs(parent_id=self._ref_doc.id,
                            granularity=self._ref_doc.granularity + 1,
                            **kwargs)
            if not chunk.mime_type:
                chunk.mime_type = self._ref_doc.mime_type
            return chunk
```

Reply:
Yes, my point is what happens if the user creates a MultiModalDocument with chunks of granularity 2. Then is not the chunks that mist have document.granularity + 1, but this new multimodaldocumet shluld have chunk.granularity - 1

Reply:
What I suggest:
1. Validate all chunks have same granularity.
2. If chunk.granularity is 0. (This means a user without caring for granularity wants to merge N documents into a multimodal document, therefore we do what u have here of increasing their granularity inside this set) Well documented please
3. If the chunk comes with granularity > 0, then it means that someone has cared to chunk already the document or that we have some driver that generates muktimodal documents in the future. Then, have document.granularity = chunk.granularity - 1. 
3. Proper testing of this
Does it make sense? What we have here does not allow for deeper levels of multimodal documents.